Fixed: "Argument 1 passed to Friendica\Util\Temporal::getRelativeDate() must be of the type string, null given"

Quix0r opened the pull request:
Changed:
- first parameter can be null

This happens when you try to add an app-specific password in 2FA settings.

Commit:
new event APIs

MrPetovan committed:
Normalize start_time and end_time parameter names in Friendica API event endpoints

tobiasd committed:
Switching to 2022.09-rc

annando committed:
Use AP data when possible

annando committed:
Improved language detection

annando committed:
Be more tolerant when receiving messages

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into performance

annando committed:
All URL fields are now binary and longer

annando committed:
Updated database.sql and messages.po

annando committed:
Updated database.sql and messages.po

annando committed:
Updated database.sql and messages.po

annando committed:
Improved logger calls

tobiasd committed:
Merge pull request #11885 from annando/tolerant-receiver

Be more tolerant when receiving messages

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into database

annando committed:
Pause the worker execution when the load is too high

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into language

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into performance

nupplaphil committed:
Merge pull request #11883 from annando/performance

Use AP data when possible

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into maxload

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into database

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into language

MrPetovan committed:
Merge pull request #11886 from annando/database

All URL fields are now binary and longer

MrPetovan committed:
Merge pull request #11884 from annando/language

Improved language detection

MrPetovan committed:
Merge pull request #11887 from annando/maxload

Pause the worker execution when the load is too high

MrPetovan committed:
Fix wrong parameter name in Core\System::httpExit

MrPetovan committed:
Enable POST route for /filerm module

- This prevents a costly page empty page render when calling the module asynchronously

MrPetovan committed:
[frio] Add support for asynchronous call to /filerm module for post categories and folders

- This prevents the page to change to an empty page when clicking on the remove link of categories and folders

annando committed:
Pagecache for frequently fetched pages

annando committed:
Issue 11890: Don't use empty contexts

annando committed:
Don't transmit empty conversations

nupplaphil committed:
Merge pull request #11893 from annando/issue-11890

Issue 11890: Don't use empty contexts

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into pagecache

annando committed:
Issue 11893: Postupdate to fix wrong parents

annando committed:
Function only needn't to be public

annando committed:
Delete the cache entry when the post is changed or deleted

annando committed:
Test stuff removed

MrPetovan committed:
Merge pull request #11894 from annando/issue-11893-a

Issue 11893: Postupdate to fix wrong parents

annando committed:
Use the cached activity function

annando committed:
Changed database version

MrPetovan committed:
Updated main translation file after moving strings

MrPetovan committed:
Merge pull request #11892 from annando/pagecache

Pagecache for frequently fetched pages

annando committed:
Smarty: Configuration added to store without sub directories

annando committed:
Read the config at another place

nupplaphil committed:
Merge pull request #11895 from annando/smarty-sub-dir

Smarty: Configuration added to store without sub directories

tobiasd committed:
HU translation updates THX Balázs Úr

nupplaphil committed:
Merge pull request #11898 from tobiasd/20220911-hu

HU translation updates THX Balázs Úr

annando committed:
Template based XML generation is replaced with native XML

nupplaphil committed:
Merge pull request #11888 from MrPetovan/bug/11861-frio-filerm

[frio] Add support for asynchronous call to /filerm module for post categories and folders

MrPetovan committed:
Merge pull request #11899 from annando/native-xml

Template based XML generation is replaced with native XML

Quix0r committed:
Changed:
- first parameter can be null

MrPetovan reviewed (changes requested):
Please address the case where it is null in the calls themselves. The `$posted_date` variable is immediately concatenated which would trip the subsequent `strtotime` call if `$posted_date` is `null`.

Quix0r commented:
Added to bug description.

annando commented:
We are deep into the RC phase. Adding type hints without checking each and every place the function is called from is problematic.

Quix0r commented:
Then let's move this over to the `develop` branch instead?

MrPetovan commented on src/Core/System.php:
```diff
@@ -468,20 +468,27 @@ public static function getLoadAvg(): array
 			DI::baseUrl()->redirect($url);
 		}
 
+		if ($includeProxyHeaders) {
+			// Send Pragma/Cache-Control headers
+			header('Pragma: no-cache');
+			header('Cache-Control: no-cache, must-revalidate');
+		}
```
What do these headers have to do with proxies?

Quix0r replied:
The team for the squid proxy has explained it in easy but accurate detail:
https://squidproxy.wordpress.com/2012/10/16/squid-3-2-pragma-cache-control-no-cache-versus-storage/

Search results:
https://searx.mxchange.org/search?q=pragma%20cache-control%20proxy&categories=general&language=en-US

MrPetovan replied:
Ok, thanks for the link. However, these headers aren't specific to proxies but any HTTP client which Squid is one. This article is informative, including about the `Pragma` header that doesn't make any sense in server responses like the ones we are generating with Friendica, so line 488 can be removed as it is effectively useless. For the other, it sounds like it should be `Cache-Control: no-store` instead. This also removes the redundant `must-revalidate` which is included in `no-cache`.

Since it isn't meant specifically for proxies, the parameter should be renamed to `$noCache` or something in this vein.

annando committed:
Detect and remove contact duplicates

MrPetovan committed:
[Composer] Update dependencies ahead of 2022.09 release

- Updating bacon/bacon-qr-code (2.0.4 => 2.0.7)
- Updating ezyang/htmlpurifier (v4.13.0 => v4.14.0)
- Updating guzzlehttp/psr7 (1.8.3 => 1.9.0)
- Updating guzzlehttp/promises (1.5.1 => 1.5.2)
- Updating symfony/polyfill-php72 (v1.23.0 => v1.26.0)
- Updating symfony/polyfill-intl-normalizer (v1.23.0 => v1.26.0)
- Updating symfony/polyfill-intl-idn (v1.23.0 => v1.26.0)
- Updating guzzlehttp/guzzle (6.5.5 => 6.5.8)
- Updating mobiledetect/mobiledetectlib (2.8.37 => 2.8.39)
- Updating monolog/monolog (1.26.1 => 1.27.1)
- Updating phpseclib/phpseclib (2.0.35 => 2.0.38)
- Updating smarty/smarty (v4.1.0 => v4.2.1)
- Updating bower-asset/vue (v2.6.14 => v2.7.10)
- Updating npm-asset/moment (2.29.1 => 2.29.4)
- Updating paragonie/constant_time_encoding (v2.4.0 => v2.6.3)
- Updating paragonie/sodium_compat (v1.17.0 => v1.18.0)
- Updating fgrosse/phpasn1 (v2.3.0 => v2.4.0)
- Updating composer/ca-bundle (1.3.1 => 1.3.3)

nupplaphil committed:
Merge pull request #11904 from MrPetovan/task/composer

[Composer] Update dependencies ahead of 2022.09 release

annando committed:
The fcontact table is now updated in the background to improve performance

annando committed:
Update src/Network/Probe.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>

annando committed:
Allow to reject specific languages on receiving posts via the relay

annando committed:
"Count" parameter added

annando committed:
Added documentation

MrPetovan committed:
Merge pull request #11905 from annando/fcontact-noscrape

The fcontact table is now updated in the background to improve performance

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into relay-languages

annando committed:
Changed documentation

annando committed:
Improved page rendering speed logging

MrPetovan committed:
Merge pull request #11908 from annando/relay-languages

Allow to reject specific languages on receiving posts via the relay

MrPetovan committed:
Merge pull request #11910 from annando/logging

Improved page rendering speed logging

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into duplicates

annando committed:
Fix forum delivery in case of beiong addressed via "@"

tobiasd committed:
Merge pull request #11914 from annando/forum-delivery

Fix forum delivery in case of being addressed via "@"

annando committed:
More cooldown calls for worker processes

Quix0r commented:
> Please address the case where it is null in the calls themselves. The `$posted_date` variable is immediately concatenated which would trip the subsequent `strtotime` call if `$posted_date` is `null`.

Within this method `$posted_date` is being checked with `is_null()` and the false result of `strtotime()`, too. So letting `null` pass here makes sense.

annando committed:
Don't perform a cooldown in the daemon by now

annando committed:
Worker for updating fcontact entries

MrPetovan committed:
Merge pull request #11902 from annando/duplicates

Detect and remove contact duplicates

annando committed:
Merge remote-tracking branch 'upstream/2022.09-rc' into cooldown

annando committed:
Reverting changes

annando committed:
Fallback for load detection, additional checks

MrPetovan commented:
The first action we perform on the parameter is a string concatenation which doesn't make sense with `null` values. I'd rather have stricter arguments declarations than relying on checks in the functions itself when the language itself can do it. If we pass a null value to this function, this means there's a potential issue in the calling code that we should investigate to avoid side-effects.

MrPetovan committed:
Merge pull request #11918 from annando/fcontact

Worker for updating fcontact entries

annando committed:
static cooldown moved to the cooldown function

MrPetovan committed:
Merge pull request #11917 from annando/cooldown

More cooldown calls for worker processes

Quix0r commented:
> The first action we perform on the parameter is a string concatenation which doesn't make sense with `null` values. I'd rather have stricter arguments declarations than relying on checks in the functions itself when the language itself can do it. If we pass a null value to this function, this means there's a potential issue in the calling code that we should investigate to avoid side-effects.

I agree with. So I better reverse my change and find the actual place and prevent this invocation from happening.

Quix0r commented:
So this bug origins while I try to add an app-specific password. A `?? ''` should be enough there then?

Quix0r commented:
Somewhere line 320 says: `if (is_null($posted_date) ...`

This has might mislead me into adding ` = null` to the parameter. So since `$posted_date` can no longer be `null`, this check will never turn `true`, too. Should I change it to `empty()` instead? Plus I can now make a unit-test for this method, should I include it in this PR?

Quix0r committed:
Changed according feedback:
- $posted_date should not allow NULL, instead use `?? ''` instead
- this happened in e.g. 2FA app-specific password listing

Quix0r commented:
I have removed the changes on the proxy-related code from this PR as this is only around `Temporal::getRelativeDate()`. Sorry, my fault mixing this.

MrPetovan commented on src/Util/Temporal.php:
```diff
@@ -317,7 +317,7 @@ public static function getRelativeDate(string $posted_date, string $format = nul
 		$abs = strtotime($localtime);
 
-		if (is_null($posted_date) || $posted_date <= DBA::NULL_DATETIME || $abs === false) {
+		if (empty($posted_date) || $posted_date <= DBA::NULL_DATETIME || $abs === false) {
 			return DI::l10n()->t('never');
 		}
```
Now that I have had a better look at it, it looks like the null value is warranted, it should return "Never" before the string concatenation and the `strtotime()` call like this:
```suggestion
	public static function getRelativeDate(string $posted_date = null, string $format = null): string
	{
		if (empty($posted_date) || $posted_date <= DBA::NULL_DATETIME) {
			return DI::l10n()->t('never');
		}
		
		$localtime = $posted_date . ' UTC';
		$abs = strtotime($localtime);
		if ($abs === false) {
			return DI::l10n()->t('never');
		}
```

Quix0r replied:
Okay, let's take a closer look. Some table columns indeed have `null` as value, e.g. "`updated`" or "`edited`" columns where on creation of a new record, they are `null` (`DATETIME NOT NULL`). So accepting `null` again makes sense.  But when I look at the method name and parameters I would expect a "please get (read: give) me a relative date from `$post_date`!" and then when I invoke it with `Temporal::getRelativeDate()` it means "please give me a relative date from nothing". Okay, that's "never" again. Hmm.

Quix0r replied:
Plus when `strtotime()` is returning `false` we interpret it as "never" while the documentation says error/bogus. Like e.g. `$bogus = strtotime('Not good');` will return `false`.

MrPetovan committed:
Ward against missing keys in Model\APContact::isRelay

- Address https://github.com/friendica/friendica/issues/11632#issuecomment-1231904280

MrPetovan committed:
Check for network key existence before calling Probe::isProbable in Model\Contact

- Address https://github.com/friendica/friendica/issues/11630#issuecomment-1253225607

Commit:
Do not remove trailing slash from URIs

annando committed:
Fix error "Class 'Friendica\Worker\Worker' not found"

annando committed:
(Hopefully) SQL improvements

annando committed:
Merge pull request #11923 from MrPetovan/bug/fatal-errors

Check for network key existence before calling Probe::isProbable in Model\Contact

annando committed:
Merge pull request #11922 from MrPetovan/bug/notices

Ward against missing keys in Model\APContact::isRelay

MrPetovan committed:
Merge pull request #11924 from annando/error

Fix error "Class 'Friendica\Worker\Worker' not found"

MrPetovan committed:
Merge pull request #11926 from annando/sql-queries

(Hopefully) SQL improvements

MrPetovan committed:
Merge pull request #11925 from mexon/mat/dont-remove-slash

Do not remove trailing slash from URIs

annando committed:
Merge pull request #11763 from tobiasd/marcor-newapi

new event APIs

annando committed:
Add the uri-id if missing

MrPetovan committed:
Merge pull request #11927 from annando/missing-uri-id

Add the uri-id if missing

MrPetovan committed:
Fix removing falsy text nodes in HTML::tagToBBCodeSub

- This wrongly removed text nodes containing just '0'

annando committed:
Merge pull request #11928 from MrPetovan/bug/11851-content-0

Fix removing falsy text nodes in HTML::tagToBBCodeSub

Quix0r commented:
I better close this ticket as it happened during the RC phase and I have already make a new branch based on current `develop` code.
